charmcraft/config.py

```diff
@@ -0,0 +1,152 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Central configuration management."""
+
+import pathlib
+from urllib.parse import urlparse
+
+import attr
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.utils import load_yaml
+
+
+def _get_key(config, field):
+    """Get the complete path of the key in the config."""
+    section = getattr(config, 'section', None)
+    if section:
+        key = '{}.{}'.format(section, field.name)
+    else:
+        key = field.name
+    return key
+
+
+def _check_type(config, field, value):
+    """Check that the value is instance of the specified type."""
+    if not isinstance(value, field.type):
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' must be a {}: got {!r}."
+            .format(key, field.type.__name__, value))
+
+
+def _check_url(config, field, value):
+    """Check that the URL has at least scheme and net location."""
+    url = urlparse(value)
+    if not url.scheme or not url.netloc:
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' must be a full URL "
+            "(e.g. 'https://some.server.com'): got {!r}.".format(key, value))
+
+
+def _check_typefield(config, field, value):
+    """Check that the value of 'type' field is valid."""
+    # None is allowed until we make mandatory this configuration field
+    if value not in ('charm', 'bundle', None):
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' (if present) must value 'charm' or 'bundle': "
+            "got {!r}.".format(key, value))
+
+
+@attr.s(kw_only=True, frozen=True)
+class _CharmhubConfig:
+    section = 'charmhub'
+
+    api_url = attr.ib(
+        type=str, default='https://api.staging.charmhub.io',
+        validator=[_check_url, _check_type])
+    storage_url = attr.ib(
+        type=str, default='https://storage.staging.snapcraftcontent.com',
+        validator=[_check_url, _check_type])
+
+    @classmethod
+    def from_dict(cls, source=None):
+        """Build from a raw dict."""
+        if source is None:
+            return cls()
+
+        if not isinstance(source, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'charmhub' field must be a dict: got {!r}."
+                .format(type(source).__name__))
+        return cls(**source)
+
+
+class _BasicPrime(tuple):
+    """Hold the list of files to include, specified under parts/bundle/prime configs.
+
+    This is a intermediate structure until we have the full Lifecycle in place.
+    """
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        if not isinstance(parts, dict):
```

Comment:
one advantage of using jsonschema here is that you eventually get autocomplete on IDEs (VS Code, PyCharms, Emacs, etc)

Reply:
Will give it a try to `jsonschema`, I think all could be simpler and more declarative when using this.